### PR TITLE
Initial (GCP-specific) implementation of Short Read Alignment pipeline.

### PR DESCRIPTION
**Pull request by gbggrant**
Initial (GCP-specific) implementation of Short Read Alignment pipeline.
Work In Progress:
- Only runs on one sample, with the lanelets for that sample organized in a specific fashion in cloud storage.
- Uses docker images built using Jon's scripts, pushed to GCR

### CODE REVIEW

File: pipelines/short-read-alignment-vector/SRA_Test1.json [hunk 1]
```diff
@@ -0,0 +1,37 @@
+{
+  "ShortReadAlignment.sample_id": "AN0131-C",
+  "ShortReadAlignment.output_basename": "SRA_Test1",
+  "ShortReadAlignment.read_group_ids": [
```
Comment (gbggrant) [inline pipelines/short-read-alignment-vector/SRA_Test1.json:4]: These were (manually, for now) pulled from the fastqs to put into the bam created by bwa mem as part of the RG string.  Not sure these are needed at all or if we only need the sample ID?

File: pipelines/short-read-alignment-vector/ShortReadAlignment.options.json [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{
+  "read_from_cache": true,
+  "write_to_cache": true,
+  "monitoring_script": "gs://broad-gotc-test-storage/cromwell_monitoring_script.sh"
```
Comment (gbggrant) [inline pipelines/short-read-alignment-vector/ShortReadAlignment.options.json:4]: FYI - this is a script that monitors mem/disk/cpu usage on google VMs in parallel with the executing task - very helpful for figuring out how much memory/disk a task needs.

File: pipelines/short-read-alignment-vector/ShortReadAlignment.wdl [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+## This is an initial proof of concept implementation.  It is designed to ONLY work on one sample, with ONLY its
+## list of input fastqs.  It is currently implemented to run using Cromwell with a google cloud platform backend.
+##
+
+workflow ShortReadAlignment {
+  String pipeline_version = "0.0.1"
+
+  input {
+    Array[String] read_group_ids
+    Array[File] input_fastq1s
```
Comment (gbggrant) [inline pipelines/short-read-alignment-vector/ShortReadAlignment.wdl:17]: These arrays were generated by taking one sample's set of inputs and pushing to the cloud.  Just structured like this for ease of (initial) testing.

File: pipelines/short-read-alignment-vector/ShortReadAlignment.wdl [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+  Float ref_size = size(reference.ref_fasta, "GiB") + size(reference.ref_fasta_index, "GiB") + size(reference.ref_dict, "GiB")
+  Float bwa_ref_size = ref_size + size(reference.ref_amb, "GiB") + size(reference.ref_ann, "GiB") + size(reference.ref_bwt, "GiB") + size(reference.ref_pac, "GiB") + size(reference.ref_sa, "GiB")
+  Float disk_multiplier = 2.5
+  Int disk_size = ceil(fastq_size + bwa_ref_size + (disk_multiplier * fastq_size) + 20)
+
+  command <<<
+    set -o pipefail
+    set -e
+
+    # TODO - need to set the read group id correctly here.
```
Comment (gbggrant) [inline pipelines/short-read-alignment-vector/ShortReadAlignment.wdl:147]: As mentioned in the comment in the json input file above.

File: pipelines/short-read-alignment-vector/ShortReadAlignment.wdl [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+    java -Xms3500m -jar /root/picard.jar \
+      MergeSamFiles \
+      INPUT=~{sep=' INPUT=' input_files} \
+      OUTPUT=~{output_filename}
+  }
+  runtime {
+    docker: runTimeSettings.picard_docker
+    preemptible: runTimeSettings.preemptible_tries
+    disks: "local-disk " + disk_size + " HDD"
+    memory: "3.75 GiB"
```
Comment (gbggrant) [inline pipelines/short-read-alignment-vector/ShortReadAlignment.wdl:216]: A note on the memory allocation strategy.  Google cloud VMs come in standard sizes (https://cloud.google.com/compute/all-pricing#n1_standard_machine_types).  The cheapest is a 1 cpu model with 3.75 G of memory (the next is 7.G w/2 cores...).  So I normally just ask for that 3.75 and in the Java invocation use -Xms to preallocate all the memory to the process.  This won't work for running on LSF.

File: pipelines/short-read-alignment-vector/ShortReadAlignment.wdl [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+
+  command {
+    java -Xms3500m -jar /root/picard.jar \
+      ValidateSamFile \
+      INPUT=~{input_file} \
+      OUTPUT=~{report_filename} \
+      REFERENCE_SEQUENCE=~{reference.ref_fasta} \
+      ~{"MAX_OUTPUT=" + max_output} \
+      IGNORE=~{default="null" sep=" IGNORE=" ignore} \
+      IGNORE=MISSING_PLATFORM_VALUE \
```
Comment (gbggrant) [inline pipelines/short-read-alignment-vector/ShortReadAlignment.wdl:364]: ValidateSamFile is failing for missing 'PL' tag - hence this line.

File: pipelines/short-read-alignment-vector/ShortReadAlignment.wdl [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+      IS_BISULFITE_SEQUENCED=false \
+      MODE=VERBOSE
+    exitCode=$?
+    if [ $exitCode != 0 ]; then
+       echo "TODO:  It failed ValidateSameFile - but I'm letting it pass for testing."
```
Comment (gbggrant) [inline pipelines/short-read-alignment-vector/ShortReadAlignment.wdl:369]: It's also failing for another reason.  To discuss.

Comment (gbggrant) [reply pipelines/short-read-alignment-vector/ShortReadAlignment.wdl:369]: Thanks @tnguyensanger I'm also hesistant to have ValidateSamFile use VALIDATION_STRINGENCY=SILENT.  I think there may be a way to have it ignore the NM flagged errors (but might be in a newer version of Picard) - will look further

File: pipelines/short-read-alignment-vector/SRA_Test1.json (hunk 1 above)
Comment (gbggrant) [reply pipelines/short-read-alignment-vector/SRA_Test1.json:4]: Thanks - I'll update that.  I think that some of these values can be pulled from the fastq with a simple grep - but if these are the sort of fields that could be put directly into the input manifest (like the sample manifest), that would be far simpler and more reliable.

File: pipelines/short-read-alignment-vector/ShortReadAlignment.wdl (hunk 7 above)
Comment (alimanfoo) [reply pipelines/short-read-alignment-vector/ShortReadAlignment.wdl:369]: > See [samtools/samtools#717 (comment)](https://github.com/samtools/samtools/issues/717#issuecomment-324943093).

Wow, that is a rabbit hole!

Comment (alimanfoo) [reply pipelines/short-read-alignment-vector/ShortReadAlignment.wdl:369]: OK, I'm happy to remove the `samtools calmd` step, if @gbggrant you can confirm it's not used by GATK UnifiedGenotyper (I'm guessing it's not, but thought I should double-check). We can then run ValidateSamFile properly.

Comment (gbggrant) [reply pipelines/short-read-alignment-vector/ShortReadAlignment.wdl:369]: @alimanfoo I have confirmed with Eric Banks that the NM (and MD) tags are not used by GATK's UnifiedGenotyper.  I'm going to take out the usage of `samtools calmd` here

Comment (alimanfoo) [reply pipelines/short-read-alignment-vector/ShortReadAlignment.wdl:369]: Perfect, thank you. I'll take it out of the spec too.

**Comment by gbggrant** [comment]
I'm merging this PR, will put the changes related to removal of calmd in another issue/PR

**Pull request merged**